### ISSUE TITLE
About the importance of PSL 

### ISSUE THREAD

**Issue by David-Lee-1990** [stasl0217/UKGE#3]
Hi, thank you for sharing your code and data. 

I am curious on the importance of PSL and I conducted serval experiments on data PPi5k provided here. I find that the lower 'self._p_psl' in models.py we chose, and we get  the higher nGCG of both linear and exp version of UKGE_rect. 

For example,  train with early stop with batch_size=1024, embedding_dim=128,  _p_psl=0.2, i get nDCG_linear = 0.951133, nDCG_exp=0.950328, if _p_psl=0,  nDCG_linear = 0.960531, nDCG_exp=0.50403

And i think this is on the contrary of the result in the paper 
 [http://web.cs.ucla.edu/~yzsun/papers/2019_AAAI_UKG.pdf](url). 

In fact, I think psl should be of some benifits as it describes the uncertainty by rules. But the results i found above make me puzzled. Can you share some light?

**Comment by stasl0217** [comment]
Hi David,
I guess the results you observed might probably be from the validation stage during training. As mentioned in our paper, experimental results reported in our paper are obtained on a mixture of unseen uncertain triples and negative triples, where PSL shows an advantage in distinguishing potential unseen positive from pure negative. In the current version of the code, for validation (not test) stage I only took triples from KG and this discrepancy may cause the difference. Maybe you can try to optimize this by combining very far entities as negative triples and use it together with the validation set. Then I believe you would be able to observe expected results with proper tuning.
FYI, just in case you would like to further explore applying PSL in your project, PPI5k is a pretty dense KG and PSL may not be able to show its full power under a high training proportion setting. After this paper, we have also observed in other experiments that PSL may show a more distinct advantage when the graph is much sparser, like 30%~50% data for training.

**Comment by David-Lee-1990** [comment]
hi,  i  need to clarify that those experiments that i conducted are of task "Relation Fact Ranking" but not "Relation Fact Classification".  So do I need to mix unseen uncertain triples and negative triples? 

 Through the codes provided, i think the  training procedure of task "Relation Fact Ranking" is training on train data  for every epoch and calculate "mse" and "mae" on validation data  and "nDCG" on test data for every 10 epochs. And I change the code to make early stop based on validation data's "mse".

**Comment by stasl0217** [comment]
Hi David,
You don't need to add negative triples to test ranking task. But it will affect your validation. See below.

I would like to clarify the validation stage. The four validation metrics are actually "mse_pos" (called "mse" in code), "mse_neg" and two ndcg. In our experiments we try to find the best balance between them, having neither of them going extreme. We have observed before that selecting rect/logi model with the best "mse_pos" may weaken its performance. The reported hyper-parameters may not be the best in this case, too. My guess is, if you early stop on the first metric, the selected model might tend to better fit the positive triple scores from KG and emphasize PSL a lot since PSL can help to detect the unseen positive. However, PSL also infers triples that do not exist in the given KG so far (unfortunately, they will be considered as negative triples in the ranking task even some of them may make sense). Emphasizing PSL too much may raise those triples high in the ranking task. Maybe only human judgment can do justice in this case, but human judgment is not very scalable. Under our experimental setting, balancing the positive and negative may be the best way to achieve the best performance.

At the same time, I would like to point out that ndcg validastion uses a randomly sampled really small subset for faster speed. So it may be very unstable. I would suggest trying a larger set for better observation.

Let me know if it helps.

**Comment by David-Lee-1990** [comment]
hi，thank you for sharing your ideas. I need to do more experiments to analysis PSL. By the way, i also observed early-stopping based on 'mse' may weaken nDCG.  What is more, i observed the larger n_psl in param.py, the better nDCG we get. 

I think the confidence got by psl sometimes conflict with the prior assumption "confidence of all unseen relation facts should be low and even zero".  I will try to first produce all the negative samples  and corresponding softlogic samples before training  and check the ratio of those negative sample that show up in both negative data and softlogic data.  i think those data make conflict between mse and psl because their confidence need to approximate 0 guided by mse but they also should be bigger than the body confidence of grounded rules.

**Comment by YaooXu** [comment]
> experimental results reported in our paper are obtained on a mixture of unseen uncertain triples and negative triples

So the final *Confidence Prediction* experimental results reported in your paper are gotten on `test_with_neg` dataset?
